### ISSUE TITLE
no (intensity) unit conversion

### ISSUE THREAD

**Issue by grimmerk** [CARTAvis/carta-archived#89]
related doc: https://docs.google.com/document/d/1VZ64s2V47mJzXX4YBBTDN6jORu9XQKc6WFy3QeCdviU/edit
*Need to consider multi-beam during conversion.* 

if the unit conversion is not frame-property dependent, such as Jy/arcsec^2 to MJy/sr, the bound values (such global min/max or the min/max per frame or clip level at non-100% levels) can be directly converted to new unit which is also not frame-property dependent. 

As long as the unit conversion dependents on frame properties, such as frequency and/or beam size (which may also change from frame to frame), conversion of all pixels needs to be done first before running percentile/quantile.

**Comment by kswang1029** [comment]
@confluence let's continue the unit related discussions here.

**Comment by confluence** [comment]
Good idea.

I am working on this in a branch now. My plan is as follows:

1. Create unit tests for the `ConversionIntensity` and `ConversionSpectral` plugins

2. Refactor and clean up the plugin code

3. Change the plugin API so that instead of accepting values, units and other data and returning converted values the plugin accepts units (and other data) and precalculates and returns a simple function / lambda which can be called repeatedly on values to be converted (possibly all the values in the image). Each conversion function which needs to be called on every frame + value pair (because the conversion depends on the frame) should ideally also be split into the portion which actually depends on the frame and any constant multiplier which can be applied once afterwards. (I.e. we should never be multiplying by a constant scaling factor inside the loop iterating over the image.)

4. Remove the current incorrect percentile calculation with location

5. Add unit awareness to the percentile calculation without location: when the percentiles are calculated for a unit which depends on the frame value, the conversion must be performed on all the values in the image before the calculation. This should be done with the function(s) returned from the plugin: a frame-dependent conversion to be called on each value individually (which may be nothing, if the conversion is not frame-dependent) and a constant multiplier which can be applied once to the final value. The percentile calculation would be performed only on intensity values; the frame value should only be used in a possible pre-conversion and then discarded. This conversion can be integrated both into the exact algorithm (where the finite values are copied to a new vector), and the approximate algorithm I plan to add (which also iterates over the image).

6. Add unit awareness to the cache: the percentiles must be recalculated and cached for every unit which cannot be derived from an existing cached unit by scaling. We should decide whether to cache only one unit per category and always scale from cached base units or whether to cache all final units.

**Comment by confluence** [comment]
I've noticed that the handling of fraction-of-peak in the plugin is asymmetrical.

The plugin can convert *from* FOP with any units for the maximum value *to* any units except FOP: this happens when the old units are specified as FOP; the actual old units after conversion from fractions are taken from the maximum value units, which makes sense.

However, if the *new* units are specified as FOP, the values are converted to fractions using the maximum value provided and keeping the existing units. No further conversion is done. There is no check in the plugin code to make sure that the units provided for the maximum value are the same as the existing units, which is potentially a bug.

I'm guessing that this behaviour doesn't mirror the conversion from FOP because it requires a maximum value to be precalculated using the new units -- which would be slow and expensive, but maybe more achievable now that there is a disk cache. Are there any plans to implement this feature, or is it not useful enough to be worth the complications?

Should the plugin implement this conversion anyway, and rely on the calling code to allow this option or not, depending on whether a new maximum is available? It's not complicated to implement in the plugin.

If it should be implemented, should we also implement a conversion from FOP to FOP in a different unit? This would require two maxima and two unit parameters to be passed to the plugin, which is impossible to do with the current API , but we could change that while we're refactoring.

**Comment by kswang1029** [comment]
can we introduce a set of new units like FOP_Jy_per_beam, FOP_Kelvin, FOP_Jy, etc? Maybe it is useful in terms of engineering and user experience (users can know that the FOP is derived with respect to what unit). This set of unit knows how it is derived, that is, the maximum in corresponding unit.

Some use cases:
1) FOP --> forbidden (no further conversion is allowed)

2) Jy/beam --> FOP_Jy_per_beam (just divide the pixels by the already found maximum)

3) Jy/beam --> FOP_Jy_per_beam --> FOP_Kelvin (the second conversion requires a unit conversion from Jy/beam to K first, then divide the pixels in Kelvin by the maximum in Kelvin)

4) Jy/beam --> Kelvin --> FOP_Jy_per_beam

5) Jy/beam --> FOP_Kelvin --> Jy/arcsec^2 (this case we do either conversion from the native Jy/beam to Jy/arcsec^2 OR we do FOP_Kelvin to Jy/arcsec^2 IF FOP_Kelvin knows how it is derived, i.e. maximum in Kelvin)

**Comment by confluence** [comment]
I assume that `Fraction of Peak` as units plus units for the maximum value is the actual format of the data in the FITS header, and we translate that directly into parameters. I'm not sure to what extent it makes sense to make custom units -- do we "rewrite" units under any other circumstances? I don't have a strong opinion about this; this is mostly an API question.

The real difficulty with conversions to FOP in a different unit is calculating the maximum: if the conversion is frame-dependent, the maximum has to be recalculated from all the values, not just scaled from the value in the old units.

**Comment by confluence** [comment]
I think the unit test for the intensity conversion plugin now covers all of the conversions supported by the existing implementation (which does not include e.g. Jy/pixel), so I'm going to proceed with the refactoring of this plugin (before moving on to do something similar with the spectral conversion plugin).

The only outstanding issue in the old code is clarity about the meaning of a `*pixels` suffix and why it is stripped off and then ignored. Is there any more information available about this?

**Comment by kswang1029** [comment]
"Fraction of Peak as units plus units for the maximum value is the actual format of the data in the FITS header, and we translate that directly into parameters." --->  I am not sure if we can see both information in standard header. BTW, don't trust DATAMAX/DATAMIN in the header. They may not be correct if users manipulate FITS data after creation.

From the users point of view, I think simply display FOP in colormap may not be good enough since after conversion, users may have no idea this FOP is with respect to what unit originally. That's why I would suggest to make custom units for this special unit "FOP". This FOP unit actually introduces significant computation cost as you mentioned for the frame-dependent case. I will mention this fact in user manual to warn users.

**Comment by kswang1029** [comment]
Regarding *pixel, I never see a unit with it. I cannot think of a physical meaning of a unit with *pixel. Effectively, it means something times a solid angle (covered by the pixel). If we are calculating total flux density, what we do is to calculate integral(I * dOmega) where I is in Jy/beam (for example) and Omega is the solid angle of the integration area. Therefore effectively here is a *pixel, but actually it will be cancelled out by the per beam.

**Comment by kswang1029** [comment]
Got a way to make a unit with *pixel! See attached image. 
To generate it:
1) open HLTau_Band7_Continuum.fits
2) switch to profiler layout
3) make a region on the image
4) go to profiler setting, profiles tab
5) change "statistic" to "sum"

Now we should see the unit changes from Jy to Jy/beam*pixel

Usually the image should be in Jy/beam and after sum (calculating Jy/beam * pixel), the unit changes to Jy/beam * pixel. For statistic="flux", unit changes from Jy/beam to Jy. For the above case (Jy -> Jy/beam * pixel), I think it is a bug.
<img width="1276" alt="screen shot 2017-08-08 at 1 13 01 pm" src="https://user-images.githubusercontent.com/20819712/29057278-c2bc9356-7c3b-11e7-8567-a8ac2852d547.png">

**Comment by confluence** [comment]
For the moment I have preserved the behaviour of the original implementation: `*pixels` is stripped off the end and ignored right at the beginning; and conversion *to* FOP is only allowed within the same units.

I suggest that we eventually implement conversion from any units to FOP in any other units, and handle the passing of the correct maximum value from outside (including disallowing certain transformations in the GUI if they would be too expensive, displaying some kind of progress indicator, etc).

But this, together with adding support for more units (like `Jy/pixel`), is not directly relevant to fixing the current issue. I brought it up here mostly to figure out how the current code is working, but there should probably be a new issue for supporting more conversions.

I have committed a new implementation of the intensity conversion plugin to my branch -- the plugin API is the same for now, but internally the conversion function delegates to another function which returns a lambda and a multiplier. Eventually I am planning to expose that function in the public API of the plugin, but I need to refactor the spectral conversion plugin in a similar way first.

Here's my updated plan:

1. Resolve any remaining issues in the intensity plugin: make the unit test error checking more strict, and possibly reorder the terms in the multiplier calculations to minimise loss of precision; make unit recognition partially case-insensitive (some real data seems to have all caps units -- but obviously the SI prefix has to be case-sensitive)

2. Create a unit test for the spectral conversion plugin internals

3. Refactor the spectral conversion plugin internals

4. Refactor the APIs of both plugins and the existing code which calls them

5. Fix unit conversion of percentiles (remove location calculation from algorithm; add pre-conversion of data to the algorithm; add awareness of units to the disk cache)

**Comment by confluence** [comment]
Followup question: how much cleaning is done on the units before they are passed into the plugin? Is the case corrected, or does the plugin get the units exactly as they appear in the FITS header?

**Comment by confluence** [comment]
In the spectral conversion plugin, kilohertz is assumed to be written as `KHz` rather than `kHz`. Is this intentional, because the unit is actually written like this in FITS files in the wild, or is this a bug?

**Comment by kswang1029** [comment]
*"Followup question: how much cleaning is done on the units before they are passed into the plugin? Is the case corrected, or does the plugin get the units exactly as they appear in the FITS header?"* ---> In the past, we tried to identify those units which can be converted to other units. For other units, like "ELECTRONS", the string from the FITS header is directly passed to colormap to show. I believe there is a normalization of convertible units (for example, Jy, JY, jy should be the same) before passing to the plugin (need to confirm). These work was done by our ex-memeber so now I cannot check with him directly.

**Comment by kswang1029** [comment]
*In the spectral conversion plugin, kilohertz is assumed to be written as KHz rather than kHz. Is this intentional, because the unit is actually written like this in FITS files in the wild, or is this a bug?* --> I think case is important and it seems to be a bug. By checking some FITS header with our image library, I feel that the case of units in FITS header is not strictly defined. I do see for example JY/BEAM and Jy/beam. For "k", since it refers to 10^3, I would keep it lower case. If possible, we may want to normalize it back to kHz before performing any other operations.  

https://fits.gsfc.nasa.gov/standard40/fits_standard40draft1.pdf <--- please check 4.3.1

Would it be possible to check how "kHz" is stored and displayed in CASA image format?

**Comment by low-sky** [comment]
We should be careful with case: mJy vs MJy

**Comment by confluence** [comment]
Yeah, since case is significant in the SI prefix we can't just do a case-insensitive match on the whole string. I'm not sure what the best solution is; possibly a specific list of allowed variants.

**Comment by confluence** [comment]
The spectral conversion plugin is implemented completely differently -- it delegates to `casacore::SpectralCoordinate` class methods which perform the conversions. The spectral coordinate object is obtained from a `casacore::CoordinateSystem` object which is constructed using the spectral index of the image -- I am investigating whether any of the conversion functions actually use the spectral index or properties derived from it, and whether the dependence on the image could be removed. Does anyone know why it may be necessary?

I should be able to make this code much less verbose than it is now, and I may be able to make the plugin follow a similar API to what I have in mind for the intensity conversion plugin (i.e. returning functions). I'm not sure yet if there is anything in the casacore functions that would make this difficult or impossible, but doing this is not as critical as it is for the intensity conversion, because it only needs to be done once per frame, not once per intensity value. We can either iterate over the image with a nested loop (over frames, and then over the image with an offset) or precalculate the values for all the frames and look them up inside a single loop, depending on which is faster.

**Comment by kswang1029** [comment]
I think only the conversion "Jy/beam <-> Kelvin" involves frequency but this should be nothing to do with spectral index. In typical FITS header, the frequency axis is defined by a reference frequency plus an increment. 

BTW (and may not be relevant to the discussion above), in casa, there is a high-level task called cl.addcomponent to allow users to generate image cubes. One usage is to set a reference frequency and a spectral index for a given image shape (e.g. point). 
Example: cl.addcomponent(dir="J2000 10h00m00.0s -30d00m00.0s", flux=[1.0,0.0,0.0,0.0], fluxunit='Jy', freq='100GHz', shape="point", polarization="Stokes", spectrumtype='spectral index', index=2.0)

**Comment by confluence** [comment]
I know that the intensity conversions are independent of the spectral index and only (sometimes) rely on frequency -- I'm asking about the internals of the spectral conversion plugin which converts between frequency, wavelength, velocity and (from what I understand) bare frame values. It currently uses the spectral index in the process of constructing the casacore object used to perform the conversions, and I'm not sure if this is necessary.

**Comment by confluence** [comment]
I have looked at the spectral conversion plugin code more, and have come to the conclusion that while it should probably be refactored and simplified at some point, refactoring is not required to fix the percentile unit conversion issue (because a much smaller number of values needs to be converted). So I'm going to shelve that for now and return to the original task of changing the API of the intensity conversion plugin and fixing the unit conversion in the percentile calculation.

**Comment by grimmerk** [comment]
@confluence , I really thank for this hard work.

When I check the lifetime of each object to using threading for new arch,  (I need to know which one is global/local/singleton), I notice that in `ViewManager.cpp`, there is a Singleton list. 

```
    Util::findSingletonObject<Clips>();
    Util::findSingletonObject<Colormaps>();
    ...
    ...
```
IMO, except for some specific purpose of a singleton (such as the cache plugin which allows everyone to set/save something in a singleton), others should be just a set of the algorithm functions (pure calculation or some getter). I notice that `UnitsIntensity` has 
```
UnitsIntensity::setDefaultUnit
UnitsIntensity::resetUnits
```
And I think it will happen that some wired/ambiguous situations, two opened file/colormaps use/share the same default unit setting? After B file is opend, it will reset this defaultUnit and A will be affected.

I'm sorry that I have not checked your current change of this pull request, so I just want to confirm that 

1. After you finish this unit conversion feature, will you use this `UnitsIntensity` file? 
2. In your design, does it allow `two open files/colormaps/users` have different unit state? (e.g. default unit)?

**Comment by confluence** [comment]
@grimmer0125 I am unfamiliar with this portion of the code, and my changes don't touch it (yet). So you can probably make any changes you need without affecting me.

I have just found a unit-related bug in the Colormap, however, while trying to figure out why spurious calls were being made to the conversion function:

The initial intensity units are never set correctly on the Colormap object despite being displayed in the GUI -- they're initialised to a default dummy value of N/A. The first time they are changed, the conversion fails and the intensity values are unchanged (because "N/A" is unrecognised), but because none of the calling functions actually *check* whether a conversion was performed, the units are set to the new units (and associated with the existing, incorrect values). Thereafter, all subsequent unit conversions are completely incorrect. A similar error occurs whenever an attempt is made to convert to/from a unit type which appears in the menu but is unsupported by the plugin (e.g. Jy/pixel, which is currently not implemented).

So there are two separate issues:

1. The conversion function needs to return a success status, and calling functions need to check it. I can do this as part of the other refactoring; the way that error handling / failure in the conversion process is currently handled is poorly defined and needs to be made more rigorous.

2. The initial unit has to be set correctly. I'm not sure if this issue has come up before, and if someone already knows how to fix it -- does #55 refer only to showing the default unit in the GUI?

**Comment by confluence** [comment]
It looks like there is an attempt to set the correct default units when the colourmap is initialised -- I am trying to debug this further to see if the problem is somewhere in `UnitsIntensity`.

I am looking more closely at `UnitsIntensity` now. I see that this class populates the menu of permitted intensity unit conversions by parsing the units from the header. However, there is a disconnect between the implementation of this class and the implementation of the conversion plugin:

1. Several units used in `UnitsIntensity` were never implemented in the plugin: `"Jy/pixel"`, `"Jy"`, `"km/s"` and `"m/s"` all appear to be possible options, but they cannot be handled by the plugin at the moment. It should be possible to add these missing conversions later; but for the moment should we not disable these unusable menu options?

2. Names are not consistent. The plugin assumes that `Kelvin` will be passed in as `"Kelvin"`, but this class assumes `"K"`. Should we support both options? Are both variants found in files?

3. `"Fraction of Peak"` is listed in the constants, but never used -- these units are not currently handled by the class (but are implemented in the plugin).

4. I have found where `"*pixels"` came from in the original plugin implementation -- it is appended to the units here, in `resetUnits`, under certain circumstances (if the data is aggregated somehow?). The function is marked as to-be-deprecated, and in the previous implementation of the plugin the suffix was just stripped and ignored. The function isn't used anywhere except the profiler.

5. In `_transTable`, if the original units match one of the five main types, optionally with an SI prefix of `M`, `m` or `u`, then the menu is populated with the original unit as-is, but then in order for each of the other four unit strings to be generated, the matching substring *including* the prefix is replaced by the canonical name of the other unit *without* the prefix. This seems like an overly elaborate way to generate an inconsistent set of units:
    * If the original units have one of these three SI prefixes or no prefix, the other units will have no prefix.
    * If the original units have one of these prefixes and an additional numeric prefix, the other units will have the same numeric prefix but no SI prefix.
    * If the original units have a different SI prefix with or without a numeric prefix, the other units will have the same prefix.

If the intention is to preserve the same prefix across all units, `[Mmu]?` should be removed from the regular expression.

**Comment by confluence** [comment]
OK, the problem is that the unit obtained from `UnitsIntensity::getDefault` is cached in the colourmap's state before the unit from the image is set via `UnitsIntensity::setDefaultUnit`. So presumably the state is not being initialised in the right order.

**Comment by confluence** [comment]
@grimmer I think the question you asked about the default unit setting is relevant here, but I don't know what the answer is in the current architecture. If there is one Colormap object shared by multiple image files, then the image-dependent state on the colourmap (both the units and the quantile values) must be updated whenever an image is selected. But this does not appear to happen (or at least I haven't been able to trigger it with a single image). What the intended behaviour is might be easier to test if I could open multiple images at once using the desktop viewer, but I haven't found a way to do this yet -- only one positional commandline parameter is used, and I can't find any UI elements for loading additional images. Am I missing something obvious?

**Comment by grimmerk** [comment]
https://github.com/CARTAvis/carta/issues/89#issuecomment-324377645

> So there are two separate issues:
> 
> The conversion function needs to return a success status, and calling functions need to check it. I can do this as part of the other refactoring; the way that error handling / failure in the conversion process is currently handled is poorly defined and needs to be made more rigorous.

Grimmer: Yes

> The initial unit has to be set correctly. I'm not sure if this issue has come up before, and if someone already knows how to fix it -- does #55 refer only to showing the default unit in the GUI?

Grimmer:
I have studied a part of the related codes but I forget something, I will try to confirm with @kswang1029 more. #55 is mainly made by EriChung (cpp side) and I modified a little on JavaScrpt side. In my memory, 

Before that fix, whatever an image file's default unit in a file header is, the UI always shows `Jy/beam`. Then EriChung tried to modify something on cpp side. His modification is 
1. open A file, shows its default unit. 
2. open B file, it A's unit is one of the compatible units for B, then colormap will keep using A's unit as used/selected unit on UI. (I'm not sure if B will try to do unit conversion or it will not do but just keep using A's unit. 
3. if A's unit is not compatible with B's unit. So just use/show B's unit on colormap. 

Then cpp side will push different states to JavaScripts. If users open A  what I see on JavaScript is, 
1. get N/A first, then get A's default unit. 
If he/she opens B later, I forget how many states JavaScript will receive. It will get its unit base on the above rules. 

Here is a compatibility table made by him. 
https://drive.google.com/open?id=1s9cY8iM4MHT3jI2VxIOVMrzTAF9MKNoN9BekOY-Whyo

https://github.com/CARTAvis/carta/issues/89#issuecomment-324445430

> 1. Several units used in UnitsIntensity were never implemented in the plugin: "Jy/pixel", "Jy", "km/s" and "m/s" all appear to be possible options, but they cannot be handled by the plugin at the moment. It should be possible to add these missing conversions later; but for the moment should we not disable these unusable menu options?

Grimmer: You mean we ignore these items and don't show them on UI? But after you finish the unit conversion plugin, they will be added back? 

> 2. Names are not consistent. The plugin assumes that Kelvin will be passed in as "Kelvin", but this class assumes "K". Should we support both options? Are both variants found in files?

I can discuss this with @kswang1029 tomorrow. I'm not sure if they are found in files, need to check more. 

4. This plugin was made by Susan and modified by EriChung, but "deprecated" is added by EriChung, as well as `_transTable`. I do not understand why profilers need to reset units. 

5. I did not check the regular expression part and I can discuss with others and check this part of the code, also this question, `preserve the same prefix across all units`

**Comment by grimmerk** [comment]
>But this does not appear to happen (or at least I haven't been able to trigger it with a single image).
Ther are some quantile/colormap sync UI issue, and one of them may be just the issue you encounter. Check English part of https://github.com/CARTAvis/carta/issues/87

The UI does not have a way to open multiple images at once. 

In my opinion, the better arch is 
1. cpp still cache/save the quantile and the unit of each file. 
2. Javascript will get and cache suitable data for UI binding. For the case, A -> B -> A, it may request the A's data or use the cached data on U side. <- When to request again can be discussed. 
3. cpp side does not use just 1 "default unit" for all images files. It can store the default unit of each file. 

The current way is to use just 1 value in `unitsIntensity` and its intention is, let `cpp manage/reflect UI state` which need to show default unit on UI. 

In the current arch, I want to remove the below situation/rule made by EriChung, it is not intuitive.
>2. open B file, it A's unit is one of the compatible units for B, then colormap will keep using A's unit as used/selected unit on UI. (I'm not sure if B will try to do unit conversion or it will not do but just keep using A's unit.

**Comment by grimmerk** [comment]
@confluence ,  In new arch of CARTA, my plan is to let each user lives in their thread/event loop and they will have their own controller, colormap and so on. But the unitIntensitiy is singleton now, so the `defaultUnit` becomes the shared & unique setting among these two users. It is very strange.

**Comment by confluence** [comment]
I have now changed the intensity conversion function in Colormap so that it returns the status -- now the units are not changed if a converter is not found, but the result (because of the initial units bug) is that no conversions are ever done (because the initial unit is `N/A` and cannot be changed).

Thanks for clarifying that only one image can be open at once; that makes a lot more sense. I have found the controls for opening a new image; what I was missing is that the viewer pane has to be selected for the option to appear in the top menu (and I also found it in the context menu).

I'm not sure why the logic for keeping / changing the unit selection menu in the colourmap needs to be complicated -- why not always regenerate the menu from the image units whenever a new image is opened? This does not appear to be an expensive task.

I have verified that the initial unit is *never* correctly set in the colourmap state -- whether the image file is opened at startup or opened from inside the program.

The solution seems to be to update the colourmap state whenever an image is loaded and the initial units are correctly set in the UI -- I am currently looking into this, but I am not familiar with this part of the architecture yet, so I'm not sure what the best way to approach it is.

My plan for caching is to cache only the base frame-dependent unit conversions of each quantile. Currently that means three values, based on the three possible conversions that will be performed on the data before the quantiles are calculated: nothing (the raw values as they are read from the file), division by the Hz value squared, and multiplication of the Hz value squared. It isn't necessary to cache conversions for every possible unit, because scaling by a constant multiplier can be performed on the calculated percentiles at the end.

I intend to pass a converter object (which contains all the necessary information) all the way down into the function in DataSource where caching decisions are made, and possibly further to whatever percentile algorithm does the calculation -- if a cached value for that frame-dependent calculation already exists, this function can use the converter to scale to the appropriate final units (I must add a helper function for this) and return the converted values. In effect this will be the same as e.g. the colourmap or the profiler performing this scaling, except that it will be delegated to a single core function.

I'm not sure how much value there is to caching the default unit for each file -- this information can easily be parsed from the header. It looks like this is already working correctly; it's just that the value is only being set in the UI and not in the object state.

I may have missed some of your questions; I will make another comment, but first I am relocating from office to home. :)

Re:

> Grimmer: You mean we ignore these items and don't show them on UI? But after you finish the unit conversion plugin, they will be added back?

Yes. Or I could start adding them now, if I can find documentation about the formulae for these conversions.

**Comment by grimmerk** [comment]
>I'm not sure why the logic for keeping / changing the unit selection menu in the colourmap needs to be complicated -- why not always regenerate the menu from the image units whenever a new image is opened? This does not appear to be an expensive task.

It is not impossible to simplify this flow but just need some time to do it (maybe just a little time). 

The current flow is 
1. launch CARTA
2. cpp generates many objects for default layout. Some objects are singleton, some are not (Controller, Colormap and so on). If we customize the layout to have 2 imageViewer, we will have two Controller objects.  Each object has its own id, e.g. c14, c15. 
3. cpp flushes partial data information into JavaScript. 
3. (embedded) browser is initialized and JavaScript engine is started. JS gets these data. 
4. The data includes layout object information (cpp defines how to layout), so JS tries to generate UI. The layout object includes its used plugin/function, e.g. Colormap. 
5. In the constructor of Colormap Window UI, it will send a request to get the corresponding Colormap function id,  then get a response including this id. This is the exact object id of Colormap object in cpp. 
6. JavaScript will use this `colormap object id` to retrieve some related object id, e.g. UnitsIntensity. 

In principle, each UI will have one or more than one "Data Object id". `the so called menu` refers to the data of Colormap (I remember colormap has another copy of embedded UnitsIntensity info.) or UnitsIntensity. 

When selecting a file, 
1. JS sends the request. 
2. JS gets the dummy response. (Most of the data are not from the direct response, are from `flushState` channel)
3. cpp will actively push several data (for different `Data object`) in different places. We call `flushState`. In `UnitsIntensity`, there is `m_state.flushState();`
4. JavaScript will get several times of `callback`, e.g. if UnitsIntensity object id is `c51` and we ever registered its (stateChange) callback, in Colormap Window related code, we will get the changed data of `UnitsIntensity`. 

`regenerate the menu` means/implies we need to ignore the previous state and just generate a new state belonging to a new image file. This change may not relate too much about the above detailed flow. And even we keep using UnitsIntensity singleton object, we still can ignore its previous state and call `flushState` to send new content of UnitsIntensity which is needed by the selected menu. 

Another way is to update the embedded UnitsIntensity information in Colormap by calling `flushState` in Colormap and modify JavaScript code to use `Colormap object information` only (I need to check). Before EriChung modified the UnitsIntensity cpp part , JavaScript would not use this itself and only use Colormap's information but cpp side would not update embedded UnitsIntensity information of Colormap object ever. 

The current usage of UnitsIntensity (for the selected unit menu) is a temporary way and we can improve it.

**Comment by grimmerk** [comment]
> The initial intensity units are never set correctly on the Colormap object despite being displayed in the GUI -- they're initialised to a default dummy value of N/A.

But the current GUI indeed show the default unit ( if we only open the 1st file). And JavaScript get two times of stateChange of some information (one is N/A, the other is not N/A), maybe you are talking about Colormap object but the GUI gets some information from UnitsIntensity (I remember but need to check)?  You can print out these two object's state before calling their `flushState`.

**Comment by confluence** [comment]
Yes, the problem is that the default unit is set correctly in the GUI, but it is not set correctly in the Colormap state. The units stored in the Colormap state are set to the default units with which the UnitsIntensity object is initialised (`N/A`) when Colormap is initialised (in `Colormap::_initializeDefaultState`) -- and this value is fetched from the state whenever the units are used (usually through `Colormap::getImageUnits` but sometimes directly). Whatever is setting the initial unit in the GUI does not update this value. I don't think it is ever updated.

When an image is loaded, the default units on the UnitsIntensity object are changed, but the value cached in the Colormap state is not updated. This is the part that needs to be added; I'm just not sure where. This needs to be handled as a change from an uninitialised value to an initialised value, not as an actual change in units, since the units aren't actually changing and the intensity values cannot (and should not) be converted.

There is a `Colormap::setImageUnits` function which seems to be linked to the UI and can be used to change the units, but it attempts to convert the bounds from the old units to the new units, which assumes that the current units are already correctly set.

**Comment by confluence** [comment]
I have added a slightly hacky fix for the initial units issue by adding some handling for this to `Colormap::_dataChanged`, which is triggered when the controller emits a `dataChanged` signal. Unfortunately this signal is used for multiple things, not just the loading of an image, so I've added some checks to attempt to determine if the units need to be updated. In the near future I think we should add a separate signal to be emitted only when a new image has been loaded.

I have also noticed that the intensity bounds are updated in this handler, even if the bounds have not changed. This will go away after the refactoring.

**Comment by confluence** [comment]
At this point I'm ready to remove the quantile calculation with location, and replace the current conversion code with something which either recalculates the quantiles or fetches the appropriate components from the cache.

In two of the calling functions (which correspond to changes in the bound percentages and changes in the selected units) part of this refactoring means storing the original percentages in the colourmap state, and passing them down to the function which may recalculate the intensity values, instead of working with saved intensity values.

However, the third calling function in Colourmap, `Colormap::setIntensityRange`, *requires* working with intensity values because it performs the reverse calculation of percentages from intensity values obtained from the GUI. ~~In this case it should be possible (and correct) to perform the conversion directly on these values, but only if the corresponding frame values are also passed in from the GUI (or if they can be deduced through other means, e.g. the currently active layer).~~ I need to look into this in more detail to see where these values are coming from.

Update: I see that these are triggered when the user enters new intensity bounds. I suspect that the correct thing to do here is to incorporate the unit conversion into the algorithm which calculates percentages from intensity values -- I will look at the algorithm to see if this makes sense.

Update: the intensity-to-percentile algorithm counts the intensity values below the given intensity. So we should be converting the image values before we compare them. We should be able to optimise this by dividing the target intensity by the constant multiplier at the start, and then applying the frame-dependent calculation on it in reverse for every frame. Then we can avoid applying the conversion to all the intensity values in the image (which can remain unchanged as we iterate over them).

**Comment by confluence** [comment]
I have noticed that there's an unnecessary round trip that happens when intensity bounds are set: after the percentage values are calculated from the intensities, intensities are calculated from those percentiles (because `Controller::applyClips` emits a `clipsChanged` signal, which triggers `Colormap::_updateIntensityBounds`).

This is probably going to contribute to performance issues on larger files. I'm not sure what the cleanest way to fix this is. We could add a boolean parameter to the functions and the signal which determines whether the intensities need to be recalculated.

In future we should perhaps also cache intensity -> percentage as well as percentage -> intensity, so that we can share the results of the calculations in both directions. There may be small differences between the results because of loss of precision, however.

**Comment by confluence** [comment]
The intensity -> percentile function is now complete.

**Comment by confluence** [comment]
The percentile -> intensity function is also complete. I have noticed odd behaviour when changing the percentiles in the image display -- the image looks very different at the same percentile depending on what the units are, and this happens even for frame-independent conversions (which should give identical results). I think I saw this happening even before the refactoring, and I assume that it is caused by converted units being used somewhere where the original image units should be used. I will try to track this down. I also need to fix some loose ends with failure / error handling.

**Comment by confluence** [comment]
Another thing: currently the spectral conversion plugin fails if there is no spectral axis (so if I understand correctly it fails on any 2D image with a single frame). Is this behaviour intentional because if there is no spectral axis specified in the file, it is impossible to determine the original units? (This plugin uses casa to do all the conversions, so whatever is possible in casa should be possible for us).

**Comment by kswang1029** [comment]
Attached is an example of single channel image from ALMA. The header defines the third axis as frequency. 
[tmp.zip](https://github.com/CARTAvis/carta/files/1265746/tmp.zip)
FYI. 

CTYPE3  = 'FREQ    '                                                            
CRVAL3  =   2.309497185811E+11                                                  
CDELT3  =   1.513671875000E+07                                                  
CRPIX3  =   1.000000000000E+00                                                  
CUNIT3  = 'Hz      '                                                            
CTYPE4  = 'STOKES  '                                                            
CRVAL4  =   1.000000000000E+00                                                  
CDELT4  =   1.000000000000E+00                                                  
CRPIX4  =   1.000000000000E+00                                                  
CUNIT4  = '        '                                                            
PV2_1   =   0.000000000000E+00                                                  
PV2_2   =   0.000000000000E+00                                                  
RESTFRQ =   2.309497185811E+11 /Rest Frequency (Hz)                             
SPECSYS = 'TOPOCENT'           /Spectral reference frame                        
ALTRVAL =  -0.000000000000E+00 /Alternate frequency reference value             
ALTRPIX =   1.000000000000E+00 /Alternate frequency reference pixel             
VELREF  =                  259 /1 LSR, 2 HEL, 3 OBS, +256 Radio    

Some 2D images (photometric observations) at other wavelengths, such as IR or optical, may not have definition on the frequency axis. So in this case, we have no information to perform spectral conversion.

**Comment by confluence** [comment]
That's not a 2D image, though; it's a 4D image with a size of 1 in two dimensions. If my understanding is correct, each CTYPE and CUNIT value corresponds to an existing axis, so if there is no third axis there will definitely be no units, and the current behaviour is intentional.

**Comment by grimmerk** [comment]
@confluence, thank you for this hard work. 
> Names are not consistent. The plugin assumes that Kelvin will be passed in as "Kelvin", but this class assumes "K". Should we support both options? Are both variants found in files?
I guess "K" in UnitsIntensity.cpp is only for UI's menu. 

So your finishing part includes ` the intensity-to-percentile algorithm counts the intensity values below the given intensity. So we should be converting the image values before we compare them.`, right?

BTW, I put 
>I have noticed that there's an unnecessary round trip that happens when intensity bounds are set: after the percentage values are calculated from the intensities, intensities are calculated from those percentiles (because Controller::applyClips emits a clipsChanged signal, which triggers Colormap::_updateIntensityBounds).

 in #102. There are many places like this and we should/will reserve a time to fix them.

**Comment by confluence** [comment]
Yes, I am converting the image values before comparing them to the intensity -- although in practice what I'm actually doing is converting the intensity, to minimise the number of required calculations. I apply the inverse of the constant multiplier, and then apply the inverse of the frame-dependent conversion for each frame. Then the individual image values don't all have to be converted -- only one calculation is required per frame (plus optionally the single constant scaling). I have also made this function process multiple inputs and outputs, so that we can iterate over the file once.

**Comment by confluence** [comment]
To summarise, outstanding issues which I think need to be fixed before my current branch can be merged (this issue could be kept open until all of the unit-related issues are resolved):

1. In the UI, the selected units shouldn't change if setting the units fails. Should we temporarily comment out all conversions in `UnitsIntensity` which are unsupported by the plugin, or should we check in the UI if the unit change failed and not change the state of the menu? (I have already made the menu not display conversions to/from Kelvin if there's no spectral axis, but this still happens with e.g. `Jy/pixel`). This is a cosmetic issue (the units in the state don't change, so there's no data corruption anymore), but it makes the UI misleading. The API function in `Colormap` which sets the units already returns an error string on failure, so something would just need to check it. This seems like a sensible thing to do even if we also remove the unsupported units.

2. The odd behaviour when units are changed and percentages are selected in the image viewer widget. I'm pretty sure this is being caused by a unit mismatch between the widgets; I'm still looking into it.

Please let me know if you notice any other issues.

**Comment by confluence** [comment]
I have partially fixed the image viewer issue -- the problem was that the intensity values being passed to the pixel pipeline are assumed to be in image units, and were being passed in in the converted units. I am now recalculating the intensities from the percentages -- but this is incorrect for frame-dependent conversions. The pixel pipeline needs to be converting the image values before comparing them to the intensity bounds (like the function which calculates the percentiles from the intensities). Has anyone looked at implementing this functionality before?

**Comment by grimmerk** [comment]
>Has anyone looked at implementing this functionality before?
@markccchiang , @appolloford  have you looked at this? 

We can comment out the unsupported conversions.  Checking the error string is another thing and may worth to do it. The question is how long and should we pop up a error UI message to alert users that the conversion fail? If all failure cases are from non-implemented conversions, we can just comment out in UnitsIntensity, but if some failures are due to unkown reasons, it is better to handle failure cases.

**Comment by confluence** [comment]
There is already an error message printed in the UI when a conversion fails. Currently I believe the only possibilities are: the conversion is unsupported, or it's frame-dependent and spectral axis is missing. I have already edited UnitsIntensity not to add conversions to/from Kelvin if there's no spectral axis, and corrected the code for not putting in conversions to/from Jy/Beam if there is no beam info. If we agree that it's best to comment out unsupported conversions for now I  can easily do that.

But the quoted text was in reference to the more serious problem, which is that there is currently no support for units in the pixel pipeline (which controls the actual display of the image for the given clipping values). For now I am passing in intensity values in the image units (recalculated from the stored percentages), but this is not the correct behaviour for frame-dependent conversions: the pixel pipeline should actually behave analogously to the algorithm for calculating percentages from intensities, and convert image values to the same units as the intensity. Or at least do the same thing as the intensity-to-percentile algorithm does, which is convert the intensity value in reverse for each frame -- but I'm not sure how easily this can be integrated into the pipeline.

The current behaviour is less buggy than the previous behaviour, but we're currently missing support for manipulating the alternate set of statistics obtained by converting the image to or from Kelvin, which seems to be an important part of being able to convert the image to those units.
